### PR TITLE
fix(inference): expose empty llm response metadata

### PR DESCRIPTION
**Pull request by lutz-grex**
## Why

When a provider returns a response without usable text, the retry helpers only log `Attempt N returned empty content` and discard everything else (`finish_reason`, blocks, `raw`, `additional_kwargs`). From the logs it is impossible to tell whether the model was blocked (safety), truncated (max tokens), returned only thinking/tool blocks, or really sent nothing. This shows up under concurrent load with some providers and is otherwise undiagnosable.

## What

Observability only. **No change to retry count, delays, error types or return values.**

- `_empty_response_diagnostics()` builds a redacted metadata dict for an unusable response: category, model, response type, `finish_reason`, block types, tool-call presence, provider request id, http status, and the *key names* of `raw` / `additional_kwargs`. Never prompt text, message content or full payloads.
- `category` gives a machine-readable reason: `provider_blocked`, `truncated`, `tool_calls_only`, `thinking_only`, `blocks_without_text`, `empty_stream`, `empty_content`, `no_response`, `no_message`.
- Reads `finish_reason` from top level, OpenAI-style `choices[0]` and Gemini-style `candidates[0]`; accepts pydantic `raw` payloads as well as dicts.
- Logged as one `WARNING` next to the existing line in all three retry paths (chat incl. stream, completion incl. stream, structured predict). Logging failures are swallowed so diagnostics can never turn a retryable response into a hard error.

Example:

```
LLM response unusable: attempt=1 diagnostics={'category': 'provider_blocked', 'stream': False, 'model': '...', 'finish_reason': 'safety', 'block_types': ['TextBlock'], 'has_tool_calls': False, 'provider_request_id': 'chatcmpl-...', 'http_status': None, 'raw_keys': ['choices', 'id', 'model'], 'additional_kwargs_keys': []}
```

## Tests

`tests/test_inference_retries.py`: parametrized category cases (empty, safety/content_filter, max_tokens/length, thinking-only, tool-calls-only, empty stream, CompletionResponse, structured `None`, pydantic raw), redaction, logger never raising, and end-to-end through all three retry helpers asserting the retry count is unchanged.

```
pytest tests     -> passed
black --check .  -> clean
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: mobilerun/agent/utils/inference.py [hunk 1]
```diff
@@ -17,6 +17,102 @@
...
+
+    def scalar(*names: str) -> object | None:
+        for src in sources:
+            for name in names:
+                if isinstance(src.get(name), (str, int, float, bool)):
+                    return src[name]
+        return None
+
+    blocks = [type(b).__name__ for b in getattr(message, "blocks", None) or []]
+    tool_calls = bool(getattr(message, "tool_calls", None) or extra.get("tool_calls"))
```
Comment (RasulOs) [inline mobilerun/agent/utils/inference.py]: codex: [P2] Read tool calls and thinking from the actual adapter fields. Anthropic stores these in `message.additional_kwargs["tool_calls"]` and `["thinking"]`; OpenAI/Gemini also use `ToolCallBlock`. The current checks miss those locations.

Confirmed against the live Anthropic API using an explicitly configured harmless native tool: the response contained `tool_use` with no text, but `acall_with_retries` logged `category="empty_content"` and `has_tool_calls=False`. No tool was executed. Please inspect the message metadata and tool-call blocks, and add regression tests using these adapter response shapes.

File: mobilerun/agent/utils/inference.py [hunk 2]
```diff
@@ -17,6 +17,102 @@
...
+
+def _empty_response_diagnostics(
+    response: object, llm: object, *, stream: bool = False
+) -> dict[str, object]:
+    """Redacted metadata (types, key names, scalars) for an unusable response. Never content."""
+    message = getattr(response, "message", None)
+    raw = _as_dict(getattr(response, "raw", None))
+    extra = _as_dict(getattr(response, "additional_kwargs", None))
+    # finish_reason lives at top level, under LiteLLM choices[0], or Gemini candidates[0]
+    sources = [raw, extra] + [
```
Comment (RasulOs) [inline mobilerun/agent/utils/inference.py]: codex: [P2] Handle the Gemini OAuth response envelope. `GeminiOAuthCodeAssistLLM` stores candidates under `raw["response"]["candidates"]`, but this extraction only checks top-level candidates.

Confirmed against live Gemini OAuth with a one-token output limit: the API returned empty text with `finishReason="MAX_TOKENS"`, but both retry attempts logged `category="empty_content"` and `finish_reason=None`. Please inspect the nested response and add a regression test asserting `category="truncated"` and `finish_reason="max_tokens"`.

**Review by RasulOs — COMMENTED** [review]
codex: Found two diagnostic classification issues, confirmed through live provider requests and the actual retry helper. Details inline.

File: mobilerun/agent/utils/inference.py [hunk 3]
```diff
@@ -17,6 +17,140 @@
...
+        or "ToolCallBlock" in blocks
+        or "tool_use" in block_types
+    )
+    thinking = bool(
+        any("think" in b.lower() for b in blocks)
+        or message_extra.get("thinking")
+        or "thinking" in block_types
+    )
+    reason = str(
+        scalar("finish_reason", "finishReason", "stop_reason", "stopReason") or ""
```
Comment (RasulOs) [inline mobilerun/agent/utils/inference.py:104]: codex: [P2] Read the OpenAI Responses incomplete reason. `MobilerunOpenAIResponses` preserves `raw.incomplete_details.reason`, but this extraction ignores it, and `_TRUNCATED` does not include `max_output_tokens`.

Confirmed with live `gpt-5.6-luna`, `max_output_tokens=16`, through the actual adapter and `acall_with_retries`: both attempts returned HTTP 200, `status="incomplete"`, `incomplete_details.reason="max_output_tokens"`, and no visible text. Both diagnostics incorrectly logged `category="thinking_only"` and `finish_reason=None`.

Please extract the incomplete reason, recognize `max_output_tokens` as truncation, and add regression tests for empty and reasoning-only Responses output. This finding concerns diagnostic classification; retry behavior is unchanged.

**Review by RasulOs — COMMENTED** [review]
codex: Confirmed one remaining diagnostic classification issue against the live OpenAI API. Details inline.

**Review by RasulOs — CHANGES_REQUESTED** [review]
